### ISSUE TITLE
[FreeBSD] HighMemoryLoadThresholdBytes change breakage

### ISSUE THREAD

**Issue by rootwyrm** [mono/mono#19017]
Observed on the FreeBSD community CI initially, caused by changes made in commit https://github.com/mono/mono/commit/ac7a77f2456e0826bfc0908b32a4df612e88304c due to FreeBSD providing a working `sysconf(3)`. Issue is at 

Specifically, FreeBSD does not provide an `_SC_AVPHYS_PAGES`; further investigation also identifies that use of `_SC_PHYS_PAGES` would not produce the desired behavior here, and the OSX method is also incompatible. At a very high level, `vm.stats.vm.v_free_count` provides an equivalent to `_SC_AVPHYS_PAGES`, but the memory management strategies are a concern.

I very much need some guidance (probably from @naricc ) as to the desired CoreCLR behavior in detail. Is the objective here to know just how many physical pages are free period, or is it to know how much total memory is _realistically_ available for use?

**Comment by akoeplinger** [comment]
@naricc

**Comment by naricc** [comment]
HighMemoryLoadThresholdBytes, in CoreCLR threshold. In CoreCLR, it is an internal number where the runtime decides that most of the physical memory is in use, at which the runtime uses more aggressive compaction strategies. It may try to reduce memory usage more aggressively in other ways too.  There is no exact equivalent thing in Mono though.  I talked to people on the CoreCLR GC Team, and the API is mostly used by C# code to decide if the memory on the machine is nearly exhausted. If the used memory is greater than this number, a software cache might decide to clear itself, for example.

Since Mono doesn’t change compaction strategies, and doesn’t really do anythign equivalent, we decided .9*physmem was a reasonable value to return.

**Comment by rootwyrm** [comment]
So if we assume that the function names are descriptive (which they do seem to be) we have a couple things to do.. for FreeBSD, we need to take `num_pages` from `vm.stats.vm.v_free_count` but we want to subtract `vm.stats.vm.v_free_min` to prevent triggering an absolute OOM.

Unfortunately, these sysctls don't have the greatest documentation. The best source tends to be D&I (McKusick, Neville-Neil, Watson, et. al) The handbook does have some basics.
https://www.freebsd.org/doc/en/books/arch-handbook/vm.html
https://www.freebsd.org/doc/en/books/design-44bsd/overview-memory-management.html
Mark Johnson also did a presentation in '17 around changes, [found here](https://papers.freebsd.org/2017/bsdtw/johnston-memory_management_in_freebsd_12.0/).

This also identifies two issues in AIX/PASE. AIX/PASE doesn't provide `sysconf(3)` in the fashion expected here of course, but this also may not handle DLPARs correctly where memory can be hot-plugged. That's going to take a bit more digging.

**Comment by rootwyrm** [comment]
So this problem is actually going to be a lot more difficult than I had expected and is going to require digging into `mono-sgen` a lot more thoroughly. It is definitely behaving very badly on FreeBSD when system memory is under load as is. The high level 'simple fix' I proposed will unquestionably break under load.

Symptomatic on 6.8.0.105 with 3 mono applications running, system somewhat but not significantly memory constrained:
```
CPU: 30.9% user,  0.0% nice, 27.2% system,  0.0% interrupt, 41.8% idle
Mem: 371M Active, 2778M Inact, 64M Laundry, 622M Wired, 383M Buf, 98M Free
Swap: 4095M Total, 4095M Free

  PID USERNAME    THR PRI NICE   SIZE    RES STATE    C   TIME    WCPU COMMAND
 8026 talecaster    7  93    0   249M   191M RUN      0 309:14 130.26% mono-sgen
 7701 talecaster   13  20    0   283M   215M ucond    0   2:07   0.18% mono-sgen

$ ps auxwww | grep mono | sort -k 5
$ ps axuwww | grep mono | sort -k 5
talecaster 8690   0.0  0.1   6928   2588  3  R+   23:14      0:00.00 grep mono
talecaster 8094   0.0  4.6 254404 189424  -  S    19:17      0:47.48 /usr/local/bin/mono --debug /usr/local/share/sonarr/NzbDrone.exe /data=/opt/talecaster/app/sonarr /nobrowser /restart (mono-sgen)
talecaster 8026 126.0  4.7 255056 195352  -  Rs   18:54    310:27.16 /usr/local/bin/mono /usr/local/share/sonarr/NzbDrone.exe --nobrowser --data=/opt/talecaster/app/sonarr (mono-sgen)
talecaster 7046   0.0  4.9 262588 203204  4  I+   14:32      1:08.68 mono app/Jackett/JackettConsole.exe -d /opt/talecaster/conf/jackett (mono-sgen)
talecaster 7701   0.0  5.3 289288 219452  2  S+   18:00      2:06.85 mono --debug ./Lidarr.exe (mono-sgen)
````

Obviously, this is not at all correct. And unfortunately, what's going on under the covers has been seen before.

```
  8026 mono-sgen RET   _umtx_op 0
  8026 mono-sgen CALL  _umtx_op(0x800c94d00,UMTX_OP_MUTEX_LOCK,0,0,0)
  8026 mono-sgen RET   _umtx_op 0
  8026 mono-sgen CALL  _umtx_op(0x800c94d00,UMTX_OP_MUTEX_LOCK,0,0,0)
  8026 mono-sgen CALL  thr_kill(0x187a5,SIGRT3)
  8026 mono-sgen RET   thr_kill 0
  8026 mono-sgen CALL  _umtx_op(0x6954b0,UMTX_OP_SEM2_WAIT,0,0,0)
  8026 mono-sgen RET   _umtx_op RESTART
  8026 mono-sgen PSIG  SIGRT3 caught handler=0x80082a850 mask=0x0 code=SI_LWP
  8026 mono-sgen CALL  sigprocmask(SIG_SETMASK,0x7fffde9f0d4c,0)
  8026 mono-sgen RET   sigprocmask 0
  8026 mono-sgen CALL  sigprocmask(SIG_BLOCK,0x7fffde9f0800,0)
  8026 mono-sgen RET   sigprocmask 0
  8026 mono-sgen CALL  _umtx_op(0x6954b0,UMTX_OP_SEM2_WAKE,0,0,0)
  8026 mono-sgen RET   _umtx_op 0
  8026 mono-sgen CALL  sigsuspend(0x7fffde9f07e8)
  8026 mono-sgen RET   _umtx_op 0
  8026 mono-sgen CALL  thr_kill(0x187a5,SIGRT2)
  8026 mono-sgen RET   thr_kill 0
  8026 mono-sgen PSIG  SIGRT2 caught handler=0x80082a850 mask=0x7ffefeff code=SI_LWP
  8026 mono-sgen CALL  _umtx_op(0x6954b0,UMTX_OP_SEM2_WAIT,0,0,0)
  8026 mono-sgen RET   sigsuspend -1 errno 4 Interrupted system call
  8026 mono-sgen CALL  sigprocmask(SIG_SETMASK,0x7fffde9f01cc,0)
  8026 mono-sgen RET   sigprocmask 0
  8026 mono-sgen CALL  sigreturn(0x7fffde9efe00)
  8026 mono-sgen RET   sigreturn JUSTRETURN
  8026 mono-sgen CALL  sigprocmask(SIG_UNBLOCK,0x695590,0)
  8026 mono-sgen RET   sigprocmask 0
  8026 mono-sgen CALL  _umtx_op(0x6954b0,UMTX_OP_SEM2_WAKE,0,0,0)
  8026 mono-sgen RET   _umtx_op 0
  8026 mono-sgen CALL  sigreturn(0x7fffde9f0980)
  8026 mono-sgen RET   sigreturn JUSTRETURN
  8026 mono-sgen CALL  _umtx_op(0x800c94d00,UMTX_OP_MUTEX_LOCK,0,0,0)
  8026 mono-sgen RET   _umtx_op 0
  8026 mono-sgen CALL  _umtx_op(0x800c94d00,UMTX_OP_MUTEX_UNLOCK,0,0,0)
```
So basically we're spinning around the drain here as fast as we possibly can. Suspends catch SIGRT2, via `pthread_kill()` which sets the whole thing spinning again, and again. So the question here is: what is mono-sgen normally trying to do here, what additional steps, what touch points are occurring internally and externally? When PHYS_MEM available is taken into play this SHOULD cause all three instances to immediately panic as we're below 0.9 (by design and by defect.)
@naricc what would be the best venue for you and I to try and walk me through this a bit more in depth? I'm hopeful there's a simple fix here still, just a different one, and I'm terrified of trying to slap a plaster on this behavior. :/

**Comment by naricc** [comment]
@rootwyrm If you want I can talk with you over the video conferencing software of your choice about this. I am on eastern standard time. What works for you? You can tag me on gitter or e-mail me at naricc@microsoft.com

**Comment by SamMonoRT** [comment]
@BrzVlad fyi
